AK/BigIntBase.h

```diff
@@ -0,0 +1,595 @@
+    }
+};
+
+// There is no reason to ban u128{0} + 1 (although the second argument type is signed, the value is
+// known at the compile time to be non-negative). In order to do so, we provide overloads in
+// UFixedBigInt which take IntegerWrapper as an argument.
+void compiletime_fail(char const*);
+
+struct IntegerWrapper {
+    StaticStorage<false, bit_width<int>> m_data;
+
+    consteval IntegerWrapper(int value)
+    {
+        if (value < 0)
+            compiletime_fail("Requested implicit conversion of an integer to the unsigned one will underflow.");
+        m_data[0] = static_cast<NativeWord>(value);
+    }
+};
+
+constexpr inline auto get_storage_of(IntegerWrapper value) { return value.m_data; }
+
+template<BuiltInUFixedInt T>
+constexpr StaticStorage<false, bit_width<T>> get_storage_of(T value)
+{
+    if constexpr (sizeof(T) > sizeof(NativeWord)) {
+        static_assert(sizeof(T) == 2 * sizeof(NativeWord));
+        return { static_cast<NativeWord>(value), static_cast<NativeWord>(value >> word_size) };
+    }
+    return { static_cast<NativeWord>(value) };
+}
+
+// ===== Utilities =====
+template<typename T>
+ALWAYS_INLINE constexpr void taint_for_optimizer(T& value)
+{
+    if (!__builtin_is_constant_evaluated()) {
+        asm volatile(""
+                     : "+rm"(value)
+                     :
+                     : "memory");
+    }
+}
+
+ALWAYS_INLINE constexpr NativeWord extend_sign(bool sign)
+{
+    return sign ? max_word : 0;
+}
+
+// FIXME: If available, we might try to use AVX2 and AVX512. We should use something similar to
+//        __builtin_ia32_addcarryx_u64 for GCC on !x86_64 or get PR79713 fixed.
+ALWAYS_INLINE constexpr NativeWord add_words(NativeWord word1, NativeWord word2, bool& carry)
+{
+    if (!__builtin_is_constant_evaluated()) {
+#if ARCH(X86_64)
+        unsigned long long output;
+        carry = __builtin_ia32_addcarryx_u64(carry, word1, word2, &output);
+        return static_cast<NativeWord>(output);
+#else
+#    if defined(AK_COMPILER_CLANG)
+        NativeWord ncarry, output;
+        if constexpr (SameAs<NativeWord, unsigned int>)
+            output = __builtin_addc(word1, word2, carry, reinterpret_cast<unsigned int*>(&ncarry));
+        else if constexpr (SameAs<NativeWord, unsigned long>)
+            output = __builtin_addcl(word1, word2, carry, reinterpret_cast<unsigned long*>(&ncarry));
+        else if constexpr (SameAs<NativeWord, unsigned long long>)
+            output = __builtin_addcll(word1, word2, carry, reinterpret_cast<unsigned long long*>(&ncarry));
+        else
+            VERIFY_NOT_REACHED();
+        carry = ncarry;
+        return output;
+#    endif
+#endif
+    }
+    // Note: This is usually too confusing for both GCC and Clang.
+    NativeWord output;
+    bool ncarry = __builtin_add_overflow(word1, word2, &output);
+    if (carry) {
+        ++output;
+        if (output == 0)
+            ncarry = true;
+    }
+    carry = ncarry;
+    return output;
```

Comment:
Is there a reason we can't use `AK::Checked<NativeWord>` here?

Reply:
Checked does not allow you to take it's value if it overflowed